### PR TITLE
Multi tenant

### LINKED ISSUE #2393: Multi-tenant uwazi core
The idea of this issue is to discuss and track all the info regarding the idea of multiple Uwazi instances running in one node.
Issues detected so far:

- [x] Sockets\
   - [x] multi node process
   - [x] multi tenant
- [x] Isolate requests per project
- [ ] Cron jobs (sync, evidence vault)
- [ ] system keys on startup ? deprecate in favor of migration if we need to add any system key.
- [x] separated databases? Possible issues with scalability
- [x] separated folders for files?
- [x] Super lengthy development 👌🤔😬 
- [x] separate db to add tenants, node process to watch changes.
- [x] Migrations should take into account multiple possible databases
_this is not being addressed, migrations will work like they do now, env variables for the db index etc and script launch._
- [ ] does semantic search and topic classification need to be aware of this ?
- [ ] tests using filesystem are a bit flaky, i am assuming due to changes on the branch to use the same testing folder (can not replicate, maybe not as serious as i thought)

### PR DESCRIPTION
**Pull request by daneryl**
fixes #2393 

## basic set up for testing multi-tenant
 - create tenants with diferent names on the shared_db, by default: 'uwazi_shared_db'
```
mongo localhost/uwazi_shared_db --eval 'db.tenants.insert({ name : "tenant1", dbName : "tenant1", indexName: "tenant1" });'
mongo localhost/uwazi_shared_db --eval 'db.tenants.insert({ name : "tenant2", dbName : "tenant2", indexName: "tenant2" });'
```

 - create databases for the newly created tenants:
```
yarn blank-state tenant1
yarn blank-state tenant2
```

- set up an nginx proxy with this basic config file
```
server {
  listen       3001;

  location / {
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_set_header tenant tenant1;
    proxy_pass http://localhost:3000;
     ## this is in case you use docker for mac 
     ## proxy_pass http://docker.for.mac.localhost:3000;
  }
}

server {
  listen       3002;

  location / {
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_set_header tenant tenant2;
    proxy_pass http://localhost:3000;
     ## this is in case you use docker for mac 
     ## proxy_pass http://docker.for.mac.localhost:3000;
  }
}

```

- start uwazi with multi_tenant activated
```
MULTI_TENANT=true yarn hot
```

now you should be able to acces tenant1 on localhost:3001 and tenant2 on 3002, tenants can be added on the fly (with the same commands above) also while uwazi is running without restarting the node process, when MULTI_TENANT is not set the instance should behave exactly like it behaves now on development for backward compatibility.

scripts like reindex and migrate still require to set up the proper env vars with the names of the db and elastic indexes.

PR checklist:
- [ ] Update READ.me ?
- [ ] Update API documentation ?

QA checklist:
- [ ] Smoke test the functionality described in the issue
- [ ] UI responsiveness
- [ ] Test for side effects
- [ ] Cross browser testing
- [ ] Code review

### CODE REVIEW

**Comment by RafaPolit** [comment]
- [ ] Check what happens on auto-migrate when the collection is behind and you start the server

**Comment by mfacar** [comment]
@daneryl, it is not possible for me export to csv, even setting the env variable TEMPORAL_FILES_FOLDER, the error I'm having is 
TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received undefined
    at validateString (internal/validators.js:121:11)
    at Object.join (path.js:1039:7)
    at temporalFilesPath (/home/mercy/projects/uwazi/uwazimt/uwazi/app/api/files/filesystem.ts:25:15)
    at /home/mercy/projects/uwazi/uwazimt/uwazi/app/api/files/exportRoutes.ts:46:64

In the development branch, the path is taking the rootpath value as default. Maybe is there any additional configuration that i need in my environment?

File: app/api/config.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  SHARED_DB: 'uwazi_shared_db',
+
+  multiTenant: process.env.MULTI_TENANT || false,
+  clusterMode: CLUSTER_MODE,
+  defaultTenant: <Tenant>{
+    dbName: process.env.DATABASE_NAME || 'uwazi_development',
+    indexName: process.env.INDEX_NAME || 'uwazi_development',
+    uploadedDocuments: UPLOADS_FOLDER || `${rootPath}/uploaded_documents/`,
+    attachments: UPLOADS_FOLDER || `${rootPath}/uploaded_documents/`,
+    customUploads: CUSTOM_UPLOADS_FOLDER || `${rootPath}/custom_uploads/`,
```
Comment (fnocetti) [inline app/api/config.ts:30]: @daneryl the default tenant is missing the temporalFiles path right?

Comment (daneryl) [reply app/api/config.ts:30]: yes it was missing, probably missed it on the rebase of the export, good catch 👍 .
it is fixed now.

**Comment by mfacar** [comment]
@daneryl, maybe is this the expected behavior? when we navigate in the same browser with different tenant in each tab, the user is emptied of the state ([example](https://www.loom.com/share/ee2a6605ceff4ce58311cc817bb0ab6c)). This is not happening with independent browsers.

**Comment by mfacar** [comment]
@daneryl, after adding a tenant on the fly, the application crashes with this message:
`uncaught exception or unhandled rejection, Node process finished !!`
`Error: tenant does not exists
    at Tenants.current (/Users/mercyfalconi/Documents/projects/uwazi/app/api/tenants/tenantContext.ts:61:13)
    at MultiTenantMongooseModel.dbForCurrentTenant (/Users/mercyfalconi/Documents/projects/uwazi/app/api/odm/MultiTenantMongooseModel.ts:20:50)`

Tenant works after restart uwazi

**Comment by daneryl** [comment]
> @daneryl, maybe is this the expected behavior? when we navigate in the same browser with different tenant in each tab, the user is emptied of the state (example). This is not happening with independent browsers.

@mfacar this is due to cookies not being [port-specific](https://stackoverflow.com/questions/44599558/is-session-port-specific) the same thing happens if you use the development branch with 2 instances in different ports, we solve this by having a proxy nginx configured with diferent domain names per instance.


> @daneryl, after adding a tenant on the fly, the application crashes with this message:
> uncaught exception or unhandled rejection, Node process finished !!
> Error: tenant does not exists at Tenants.current (/Users/mercyfalconi/Documents/projects/uwazi/app/api/tenants/tenantContext.ts:61:13) at MultiTenantMongooseModel.dbForCurrentTenant (/Users/mercyfalconi/Documents/projects/uwazi/app/api/odm/MultiTenantMongooseModel.ts:20:50)

@mfacar sorry i forgot, to be able to add tenants on the fly mongo should be running as a [replica set](https://docs.mongodb.com/manual/tutorial/deploy-replica-set/), quick replica set locally:
```
mongod --replSet rs0
mongo localhost --eval 'rs.initiate()'
```

with that, you should be able to add a tenant on the fly.
It is true that the instance is crashing when requesting for a tenant that does not exists, seems to easy to take down and i think the uwazi process should be resilient to this (im working on this), there is 2 cases that i know of that crash the process:

- tenant does not exists
- tenant exists but db for tenant does not

File: app/api/i18n/specs/routes.spec.js [hunk 2]
```diff
@@ -4,6 +4,10 @@ import i18nRoutes from 'api/i18n/routes.js';
 import instrumentRoutes from 'api/utils/instrumentRoutes';
 import translations from 'api/i18n/translations';
 
+const mocketSocketIo = () => ({
```
Comment (daneryl) [inline app/api/i18n/specs/routes.spec.js:7]: mocket ? xd

File: app/api/migrations/migrate.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  });
+  //@ts-ignore
+  errorLog.closeGraylog();
+  await DB.disconnect();
+};
+
+run().catch(async e => {
+  await DB.disconnect();
+  throw e;
+});
```
Comment (daneryl) [inline app/api/migrations/migrate.ts:22]: we should use unhandledRejection event or this exits with exit code 1 ?

File: app/api/socketio/setupSockets.ts [hunk 4]
```diff
@@ -11,10 +16,37 @@ declare global {
...
+  }
 }
 
-export default (server: Server, app: Application) => {
+const setupSockets = (server: Server, app: Application) => {
   const io: SocketIoServer = socketIo(server);
+
+  io.on('connection', socket => {
+    socket.join(socket.request.headers.tenant || tenants.defaultTenantName);
+  });
```
Comment (daneryl) [inline app/api/socketio/setupSockets.ts:31]: what happens to sockets that disconnect ?

File: app/api/sync/specs/uploadRoute.spec.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      const app = express();
+      app.use(multitenantMiddleware);
+      syncRoutes(app);
+
+      const response = await requestAPI(app)
+        .post('/api/sync/upload')
+        .set('X-Requested-With', 'XMLHttpRequest')
+        .attach('file', path.join(__dirname, 'testUpload.txt'));
+
+      const properlyUploaded = fs.existsSync(uploadsPath('testUpload.txt'));
```
Comment (daneryl) [inline app/api/sync/specs/uploadRoute.spec.ts:45]: avoid existsSync

File: app/api/utils/specs/staticFilesMiddleware.spec.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import request from 'supertest';
+import express, { Application } from 'express';
+import { setupTestUploadedPaths, uploadsPath, writeFile } from 'api/files';
+import { staticFilesMiddleware } from '../staticFilesMiddleware';
+import { testingTenants } from '../testingTenants';
+
+describe('multitenant middleware', () => {
```
Comment (daneryl) [inline app/api/utils/specs/staticFilesMiddleware.spec.ts:7]: change describe, seems copy pasted from other spec

File: app/api/utils/specs/staticFilesMiddleware.spec.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  beforeEach(() => {
+    testingTenants.mockCurrentTenant({ name: 'default' });
+    setupTestUploadedPaths();
+  });
+
+  it('should execute next middlewares inside a tenant async context', async () => {
```
Comment (daneryl) [inline app/api/utils/specs/staticFilesMiddleware.spec.ts:13]: change description, seems copy pasted from other spec

**Comment by fnocetti** [comment]
@daneryl Uwazi seems to be ignoring the MULTI_TENANT flag, even when starting up with `MUTLI_TENANT=false yarn hot` it is working on multitenant mode (I can access tenant specific instances, e.g. localhost:3001)

**Comment by daneryl** [comment]
> @daneryl Uwazi seems to be ignoring the MULTI_TENANT flag, even when starting up with `MUTLI_TENANT=false yarn hot` it is working on multitenant mode (I can access tenant specific instances, e.g. localhost:3001)

this is true, the initial idea is that there is no single-tenant mode, uwazi is always multi-tenant, the reason for the flag to exist is to temporary have some features that can not work when using uwazi as multi-tenant, basically this flag is to remove backward compatibility on some features, maybe we should rename the flag ?

File: app/api/migrations/migrate.ts (hunk 3 above)
Comment (daneryl) [reply app/api/migrations/migrate.ts:26]: this was not ending with code 1, fixed now.

File: app/api/socketio/setupSockets.ts (hunk 4 above)
Comment (daneryl) [reply app/api/socketio/setupSockets.ts:31]: the socket leaves the room on disconnection.

**Pull request merged**